### PR TITLE
Add PAL-based backend for AMD GPUs

### PR DESCRIPTION
**Pull request by FabianWildgrube**
This adds the necessary AnyDSL runtime code to interface with AMD GPUs through the PAL driver library. This poses an alternative to the HSA (ROCm) based support for AMD GPUs currently available in AnyDSL.

Depends on AnyDSL/thorin#154 in thorin and AnyDSL/anydsl#17 in the metaproject.

This PR includes work by Richard Membarth (@richardmembarth) and Felix Kawala (@FelixKawala).

### CODE REVIEW

File: platforms/artic/intrinsics_thorin.impala [hunk 1]
```diff
@@ -15,7 +15,8 @@
 #[import(cc = "thorin")] fn cuda(_dev: i32, _grid: (i32, i32, i32), _block: (i32, i32, i32), _body: fn() -> ()) -> ();
 #[import(cc = "thorin")] fn nvvm(_dev: i32, _grid: (i32, i32, i32), _block: (i32, i32, i32), _body: fn() -> ()) -> ();
 #[import(cc = "thorin")] fn opencl(_dev: i32, _grid: (i32, i32, i32), _block: (i32, i32, i32), _body: fn() -> ()) -> ();
-#[import(cc = "thorin")] fn amdgpu(_dev: i32, _grid: (i32, i32, i32), _block: (i32, i32, i32), _body: fn() -> ()) -> ();
+#[import(cc = "thorin")] fn amdgpu_hsa(_dev: i32, _grid: (i32, i32, i32), _block: (i32, i32, i32), _body: fn() -> ()) -> ();
+#[import(cc = "thorin")] fn amdgpu_pal(_dev: i32, _grid: (i32, i32, i32), _block: (i32, i32, i32), _body: fn() -> ()) -> ();
```
Comment (Hugobros3) [inline platforms/artic/intrinsics_thorin.impala:19]: This is a breaking change. Are we cool with that @richardmembarth ?

File: src/hsa_platform.h [hunk 2]
```diff
@@ -9,8 +9,8 @@
 #include <unordered_map>
 #include <vector>
 
-#include <hsa.h>
-#include <hsa_ext_amd.h>
+#include <hsa/hsa.h>
+#include <hsa/hsa_ext_amd.h>
```
Comment (Hugobros3) [inline src/hsa_platform.h:13]: is this change relevant to the PAL backend ?

File: src/pal/pal_utils.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+// Returns Pal::GpuHeap::GpuHeapCount if no appropriate heap can be found.
+Pal::GpuHeap find_gpu_local_heap(const Pal::IDevice* device, Pal::gpusize memory_size);
+
+bool allocation_is_host_visible(Pal::IGpuMemory* gpu_allocation);
+
+llvm::MDNode* get_metadata_mdnode(const llvm::Function* func, const char* key, int index = 0);
+llvm::StringRef get_metadata_string(const llvm::Function* func, const char* key);
+uint64_t get_metadata_uint(const llvm::Function* func, const char* key, int index = 0);
+
+extern const char* ComputeShaderMainFnName;
```
Comment (Hugobros3) [inline src/pal/pal_utils.h:56]: Why is this `extern` ?

File: src/pal_platform.h [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+#include <llvm/IR/DataLayout.h>
+#include <llvm/IR/LLVMContext.h>
+#include <llvm/Linker/Linker.h>
+#include <llvm/Pass.h>
+#include <llvm/Passes/OptimizationLevel.h>
+#include <llvm/Support/SourceMgr.h>
+#include <llvm/Target/TargetMachine.h>
+#endif
+
+namespace llvm_utils {
```
Comment (Hugobros3) [inline src/pal_platform.h]: This header is exposed to the rest of the runtime, but the `llvm_utils` namespace is only used inside `pal_platform.cpp`

File: src/pal/pal_fix_calling_convention_pass.h [hunk 5]
```diff
@@ -0,0 +1,14 @@
+// Copyright (c) 2023 Advanced Micro Devices, Inc. All rights reserved.
+
+#ifndef PAL_PLATFORM_FIX_CALLING_CONVENTION_H
+#define PAL_PLATFORM_FIX_CALLING_CONVENTION_H
+
+#include <llvm/IR/PassManager.h>
+
+/// This pass sets the calling convention to AMDGPU_Gfx for all calls in the given module and sets the AlwaysInline
+/// Attribute on every called function in the module to avoid the LLVM AMDGPU backend throwing errors.
```
Comment (Hugobros3) [inline src/pal/pal_fix_calling_convention_pass.h:7]: This is because PAL doesn't support a call stack, right ? But can PAL support (stackless) continuations, ie one-way jumping to another function ?

File: src/pal/pal_insert_halt_pass.h [hunk 6]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) 2023 Advanced Micro Devices, Inc. All rights reserved.
+
+#ifndef PAL_PLATFORM_INSERT_HALT_PASS_H
+#define PAL_PLATFORM_INSERT_HALT_PASS_H
+
+#include <llvm/IR/PassManager.h>
+#include <llvm/Pass.h>
+
+/// Pass that inserts RDNA specific assembly to halt a shader as soon as it starts if the environment variable
+/// "HALT_IMMEDIATELY" is set to the value "ON"
```
Comment (Hugobros3) [inline src/pal/pal_insert_halt_pass.h:8]: What is the purpose of this exactly ? Debugging GPU launch issues ?

File: src/pal/pal_lower_builtins_pass.h [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+#include <llvm/IR/PassManager.h>
+#include <llvm/Pass.h>
+#include <llvm/Transforms/IPO/PassManagerBuilder.h>
+
+#include <pal.h>
+#include <palDevice.h>
+
+/// This pass takes care of replacing calls to so-called "builtins" (i.e. local/global thread indices, and similar
+/// compute shader builtin values) with the appropriate amdgpu inline assembly that extracts the values from
+/// prepopulated SGPRs according to the RDNA2 or RDNA3 ABI. This pass only supports gfx-levels 10 and 11.
```
Comment (Hugobros3) [inline src/pal/pal_lower_builtins_pass.h:15]: Why is this done here instead of in the LLVM back-end ?

File: src/pal/pal_device.cpp [hunk 8]
```diff
@@ -0,0 +1,541 @@
...
+    // as it's not used anymore.
+    return device_->AddGpuMemoryReferences(1, &mem_ref, nullptr, Pal::GpuMemoryRefCantTrim);
+}
+
+PalDevice::GpuVirtAddr_t PalDevice::allocate_shared_virtual_memory(Pal::gpusize sizeInBytes) {
+    Pal::SvmGpuMemoryCreateInfo create_info = {};
+    create_info.size = sizeInBytes;
+    create_info.alignment =
+        4096; // This seems to be the minimum alignment needed to avoid triggering asserts in PAL
+    create_info.mallPolicy = Pal::GpuMemMallPolicy::Always;
```
Comment (Hugobros3) [inline src/pal/pal_device.cpp:281]: for my curiosity: what does "mall policy" mean ?

File: src/pal/pal_device.cpp [hunk 9]
```diff
@@ -0,0 +1,541 @@
...
+    const GpuVirtAddr_t gpu_address = memory->Desc().gpuVirtAddr;
+    assert(memory_objects_.find(gpu_address) == memory_objects_.end()
+           && "Virtual address is already in memory_objects map.");
+    memory_objects_[gpu_address] = memory;
+    return gpu_address;
+}
+
+void PalDevice::forget_memory(GpuVirtAddr_t gpu_address) {
+    assert(memory_objects_.find(gpu_address) != memory_objects_.end()
+           && "Virtual address is already in memory_objects map.");
```
Comment (Hugobros3) [inline src/pal/pal_device.cpp]: wrong error message ?

File: src/pal/pal_utils.h [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+} // namespace pal_utils
+
+#define CHECK_PAL(err, name) pal_utils::check_pal_error(err, name, __FILE__, __LINE__)
```
Comment (Hugobros3) [inline src/pal/pal_utils.h]: Nit: I prefer this style: https://github.com/shady-gang/shady/blob/master/src/runtime/vulkan/vk_runtime_private.h#L65, removes the duplication and associated problems (inconsistency, typos, ...)

File: src/pal_platform.h [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+// Initializes llvm with amdgpu using the following commandline options:
```
Comment (Hugobros3) [inline src/pal_platform.h]: Where is the envvar set to this ? I don't see a matching change in the metarepo.

File: src/pal/pal_device.cpp [hunk 12]
```diff
@@ -0,0 +1,541 @@
...
+    result = device_->CreateGpuMemory(create_info, malloc(gpu_memory_size), gpu_memory_pp);
+    CHECK_PAL(result, "CreateGpuMemory()");
+
+    Pal::GpuMemoryRef mem_ref = {};
+    mem_ref.pGpuMemory = *gpu_memory_pp;
+
+    // Make this allocation permanently resident. We might want to track  which allocations are actually being
+    // used and only make those resident but in AnyDSL usually the user takes care of releasing memory as soon
+    // as it's not used anymore.
+    return device_->AddGpuMemoryReferences(1, &mem_ref, nullptr, Pal::GpuMemoryRefCantTrim);
```
Comment (Hugobros3) [inline src/pal/pal_device.cpp:273]: Tracking allocations like that is probably unworkable since there's no concept of binding in AnyDSL's runtime API, we do everything with buffer addresses or handles that we can provide through arbitrary channels.

Yes - the CL backend will bind buffers that are passed to the main function, but that's because old-school CL stinks ! I'm fairly sure we don't want that going forward.

If GPU memory pressure proves a real issue we can always expose the concept of residency.

File: src/pal/pal_lower_kernel_arguments_pass.h [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+#define PAL_PLATFORM_LOWER_KERNEL_ARGUMENTS_H
+
+#include <llvm/IR/PassManager.h>
+
+/// This pass replaces accesses to kernel arguments with loads from offsets from a manually supplied buffer
+/// containing these arguments. The pointer to this buffer is expected to be prepopulated into specific sgprs
+/// by the PALPlatform.
+///
+/// This pass is an almost 1:1 replicate of the AMDGPULowerKernelArguments pass
+/// (llvm-project/llvm/lib/Target/AMDGPU/AMDGPULowerKernelArguments.cpp)
```
Comment (Hugobros3) [inline src/pal/pal_lower_kernel_arguments_pass.h:11]: Is there a reason why we wouldn't just use the LLVM pass?

File: src/hsa_platform.h (hunk 2 above)
Comment (FabianWildgrube) [reply src/hsa_platform.h:13]: I don't think so. This was added by @richardmembarth, though.

File: src/pal/pal_utils.h (hunk 3 above)
Comment (FabianWildgrube) [reply src/pal/pal_utils.h:56]: It's declared here, defined in the accompanying `.cpp` file and used in the various llvm passes needed for amdpal jit compilation, so it needs external linkage.

File: src/hsa_platform.h (hunk 2 above)
Comment (richardmembarth) [reply src/hsa_platform.h:13]: Recent HSA/ROCm installations changed the header path.

File: src/pal_platform.h (hunk 4 above)
Comment (FabianWildgrube) [reply src/pal_platform.h]: Good point. I'll remove it from the header.

File: src/pal/pal_fix_calling_convention_pass.h (hunk 5 above)
Comment (FabianWildgrube) [reply src/pal/pal_fix_calling_convention_pass.h:7]: This is not a PAL thing, but rather a LLVM AMDGPU backend thing. Jumps are supported in theory but we ran into issues with the ocml and other library functions. To be on the safe side we force-inline everything in the IR and then the backend is happy. Since our focus was to get things up and running we didn't spend more time on this problem yet.

File: src/pal/pal_insert_halt_pass.h (hunk 6 above)
Comment (FabianWildgrube) [reply src/pal/pal_insert_halt_pass.h:8]: No, it's used to enable inspecting SGPR values set during work group launch on the GPU right at the beginning of a shader program before any ISA instructions are executed.

File: src/pal/pal_lower_builtins_pass.h (hunk 7 above)
Comment (FabianWildgrube) [reply src/pal/pal_lower_builtins_pass.h:15]: Mainly flexibility. These builtins rely on SGPR set up and we control that setup explicitly through `pal_utils::build_metadata()`. If we want to change things up, we can do so because we're not necessarily bound to the specific ABI implemented by the LLVM back-end.
I see the point, tough. We could "out-source" at least work item id and work group id builtins to the LLVM back-end.

File: src/pal/pal_device.cpp (hunk 8 above)
Comment (FabianWildgrube) [reply src/pal/pal_device.cpp:281]: MALL is the last level cache on AMD GPUs. You can define for each allocation whether it should bypass the last-level cache or not.

File: src/pal/pal_device.cpp (hunk 12 above)
Comment (FabianWildgrube) [reply src/pal/pal_device.cpp:273]: Alright :+1: I'll adjust the comment to only state that the allocation is marked as resident until it's released explicitly by the user program.

File: src/pal/pal_lower_kernel_arguments_pass.h (hunk 13 above)
Comment (FabianWildgrube) [reply src/pal/pal_lower_kernel_arguments_pass.h:11]: Yes. The LLVM pass relies on `gcn_dispatch_ptr()`, which relies on the convention that every HSA dispatch is triggered by an AQL packet that provides a buffer with the kernel arguments. We don't have all of that luxury with PAL out of the box, so we build up the buffer with the kernel arguments manually. The problem with this is, that the pointer to that buffer is placed in SGPRs, which is not what the LLVM pass expects. So, the change we had to make is the code that's generated to get the pointer to the kernargs buffer. Since this is "buried" inside the function of the original LLVM pass I didn't see a better option than moving the pass here and adding our modifications. (See comments in the cpp file)

File: src/pal/pal_device.cpp (hunk 9 above)
Comment (FabianWildgrube) [reply src/pal/pal_device.cpp]: Yes, good catch! :)

File: src/pal_platform.h (hunk 11 above)
Comment (FabianWildgrube) [reply src/pal_platform.h]: Hmm. I think we may have missinterpreted the comment in HSAPlatform::emit_gcn():
https://github.com/AnyDSL/runtime/blob/36a5f3b7bc763497c22a7053230ea8a1012329b5/src/hsa_platform.cpp#L575
Is it actually just in there to set those variables quickly?

Comment (richardmembarth) [reply src/pal_platform.h]: You can set those code-gen variables via the environment variable `ANYDSL_LLVM_ARGS`, see:
https://anydsl.github.io/Device-Code-Generation-and-Execution.html#amdgpu-code-generation-optimization
They have been just added as a comment to the code.

Comment (FabianWildgrube) [reply src/pal_platform.h]: I see :+1: I've fixed the comment to reflect that better.

File: platforms/artic/intrinsics_thorin.impala (hunk 1 above)
Comment (richardmembarth) [reply platforms/artic/intrinsics_thorin.impala:19]: This is fine from my side, unless we find a more elegant way (like adding a default argument that specifies for which API we want to generate code for).
